### PR TITLE
Add unit testing for pinterest Authenticator  and update Documentation

### PR DESCRIPTION
**Pull request by nalakas**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+        return new Object[][]{{authenticatorProperties}};
+    }
+
+    @BeforeMethod
+    public void setUp() {
+        pinterestAuthenticator = new PinterestAuthenticator();
+        initMocks(this);
+    }
+
+    @Test
```
Comment (menakaj) [inline component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: Add test descriptions to the missing test cases.

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+            dataProvider = "getAuthenticatorPropertiesData")
+    public void testProcessAuthenticationResponse(Map<String, String> authenticatorProperties) throws Exception {
+        PinterestAuthenticator spyAuthenticator = PowerMockito.spy(new PinterestAuthenticator());
+        PowerMockito.when(httpServletRequest.getParameter(anyString())).thenReturn("method");
+        context.setAuthenticatorProperties(authenticatorProperties);
+        PowerMockito.mockStatic(OAuthAuthzResponse.class);
+        Mockito.when(OAuthAuthzResponse.oauthCodeAuthzResponse(Mockito.any(HttpServletRequest.class))).thenReturn(authAuthzResponse);
+        PowerMockito.mockStatic(OAuthClientRequest.class);
+        //PowerMockito.doReturn(authAuthzResponse).when(mockOAuthAuthzResponse," oauthCodeAuthzResponse",httpServletRequest);
+        Mockito.when(OAuthClientRequest.tokenLocation(Mockito.anyString())).thenReturn(new OAuthClientRequest.TokenRequestBuilder("https://api.pinterest.com/v1/oauth/token"));
```
Comment (menakaj) [inline component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: Styling issue. Exceeds 120 character limit.

Comment (menakaj) [reply component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: Find and fix in other places as well.

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+        PowerMockito.whenNew(OAuthClient.class).withAnyArguments().thenReturn(oAuthClient);
+        Mockito.when(mockOAuthClient.accessToken(mockOAuthClientRequest)).thenReturn(oAuthJSONAccessTokenResponse);
+        Mockito.when(oAuthClient.accessToken(Mockito.any(OAuthClientRequest.class))).thenReturn(oAuthJSONAccessTokenResponse);
+        spyAuthenticator.processAuthenticationResponse(httpServletRequest, httpServletResponse, context);
+    }
+
+    @Test(expectedExceptions = InvalidCredentialsException.class, description = "Negative test case for HandleErrorResponse")
+    public void testHandleErrorResponse() throws Exception {
+        Whitebox.invokeMethod(pinterestAuthenticator, "handleErrorResponse", httpServletRequest);
+
```
Comment (menakaj) [inline component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: Remove unnecessary empty lines.

File: component/src/test/resources/testng.xml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  -->
+
+<suite name="Suite1" verbose="1" >
+
+    <test name="pinterest-Authenticator-Unit-Tests">
+        <classes>
+            <class name="org.wso2.carbon.identity.authenticator.pinterest.PinterestTest"/>
+        </classes>
+    </test>
+</suite>
```
Comment (menakaj) [inline component/src/test/resources/testng.xml]: Add a new line at the end of the file.

File: docs/sampleApp.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    echo "modules/samples/" >> sparse-checkout
+    ```
+    
+* Navigate out of .git/info directory and checkout the v5.4.0 tag to update the empty repository with the remote one.
+    ```
+    cd ..
+    cd ..
+    git checkout -b v5.4.0 v5.4.0
+    ```
+* Go to is-samples/modules/samples/sso/sso-agent-sample directory and run `mvn clean install` and get the war file from the target folder.
```
Comment (menakaj) [inline docs/sampleApp.md]: It is better if we can say the name of the war file.

File: pom.xml [hunk 6]
```diff
@@ -339,5 +400,13 @@
...
+        <log4j.version>1.2.13</log4j.version>
+        <!--Test Dependencies-->
+        <junit.version>4.12</junit.version>
+        <testng.version>6.9.10</testng.version>
+        <jacoco.version>0.7.9</jacoco.version>
+        <powermock.version>1.6.5</powermock.version>
+        <maven.surefire.plugin.version>2.18.1</maven.surefire.plugin.version>
+        <slf4j.api.version>1.7.12</slf4j.api.version>
     </properties>
 </project>
```
Comment (menakaj) [inline pom.xml]: Add a new line.

**Review by menakaj — CHANGES_REQUESTED** [review]

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java (hunk 1 above)
Comment (nalakas) [reply component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: fixed

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java (hunk 2 above)
Comment (nalakas) [reply component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: fixed in other places as well

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java (hunk 3 above)
Comment (nalakas) [reply component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: fixed

File: component/src/test/resources/testng.xml (hunk 4 above)
Comment (nalakas) [reply component/src/test/resources/testng.xml]: line added

File: docs/sampleApp.md (hunk 5 above)
Comment (nalakas) [reply docs/sampleApp.md]: improved

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
+package org.wso2.carbon.identity.authenticator.pinterest;
```
Comment (kesavany) [inline component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: Add License header.

File: component/src/test/resources/testng.xml [hunk 8]
```diff
@@ -0,0 +1,27 @@
+<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >
+<!--
+  ~ Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (kesavany) [inline component/src/test/resources/testng.xml]: Change year in the License header.

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java (hunk 7 above)
Comment (nalakas) [reply component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: fixed

File: component/src/test/resources/testng.xml (hunk 8 above)
Comment (nalakas) [reply component/src/test/resources/testng.xml]: fixed

**Review by menakaj — APPROVED** [review]

File: component/pom.xml [hunk 9]
```diff
@@ -93,6 +93,119 @@
...
+                    <artifactId>jcl-over-slf4j</artifactId>
+                </exclusion>
+                <exclusion>
+                    <groupId>org.slf4j</groupId>
+                    <artifactId>log4j-over-slf4j</artifactId>
+                </exclusion>
+            </exclusions>
+        </dependency>
+    </dependencies>
 </project>
```
Comment (praminda) [inline component/pom.xml]: Put a empty new line at the end of file.
```diff
-</project> 
+</project>
+

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+import java.io.ByteArrayInputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.util.HashMap;
+import java.util.Map;
+
+import static org.mockito.MockitoAnnotations.initMocks;
+
+import org.wso2.carbon.identity.application.common.model.ClaimMapping;
```
Comment (praminda) [inline component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: Import is not sorted properly

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+
+@RunWith(PowerMockRunner.class)
+@PrepareForTest({OAuthAuthzResponse.class, OAuthClientRequest.class})
+public class PinterestTest extends PowerMockTestCase {
+
+    @ObjectFactory
+    public IObjectFactory getObjectFactory() {
```
Comment (praminda) [inline component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: Move the method below the property declarations.

Comment (nalakas) [reply component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: fiexed in https://github.com/wso2-extensions/identity-outbound-auth-pinterest/pull/8

File: component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java (hunk 10 above)
Comment (nalakas) [reply component/src/test/java/org/wso2/carbon/identity/authenticator/pinterest/PinterestTest.java]: fiexed in https://github.com/wso2-extensions/identity-outbound-auth-pinterest/pull/8

File: component/pom.xml (hunk 9 above)
Comment (nalakas) [reply component/pom.xml]: fiexed in https://github.com/wso2-extensions/identity-outbound-auth-pinterest/pull/8

**Pull request merged**